// react-switch-selector/dist/index.modern.js
import e,{createElement as o,useEffect as t}from"react";import{setup as r,styled as n}from"goober";import{shouldForwardProp as i}from"goober/should-forward-prop";function d(){return(d=Object.assign||function(e){for(var o=1;o<arguments.length;o++){var t=arguments[o];for(var r in t)Object.prototype.hasOwnProperty.call(t,r)&&(e[r]=t[r])}return e}).apply(this,arguments)}let l,a,s,c,p=e=>e;const u=["fontSize","fontColor","selectedFontColor","optionsAmount","optionBorderRadius","isRawText","selectedIndex","border","backgroundColor","selectedBackgroundColor","wrapperBorderRadius","selectionIndicatorMargin","disabled"];r(o,void 0,void 0,i(e=>!u.includes(e)));const g=n("div")(l||(l=p`
    display: flex;
    border-radius: ${0}px;
    border: ${0};
    background: ${0};
    width: 100%;
    height: 100%;
    position: relative;
    opacity: ${0};

    &::before {
        top: 50%;
        left: ${0}%;
        content: "";
        position: absolute;
        height: calc(100% - ${0}px);
        width: calc(
            ${0}%${0}${0}px
        );
        border-radius: ${0}px;
        border: ${0}px solid
            ${0};
        background: ${0};
        transition: left 0.1s linear, background 0.1s linear;
        transform: translateY(-50%);
        z-index: 1;
        box-sizing: content-box;
    }
`),e=>e.wrapperBorderRadius,e=>e.border,e=>e.backgroundColor,e=>e.disabled?.7:1,e=>e.selectedIndex/e.optionsAmount*100,e=>2*e.selectionIndicatorMargin,e=>1/e.optionsAmount*100," - ",e=>2*e.selectionIndicatorMargin,e=>e.optionBorderRadius,e=>e.selectionIndicatorMargin,e=>e.backgroundColor,e=>e.selectedBackgroundColor),b=n("div")(a||(a=p`
    display: flex;
    align-items: center;
    height: 100%;
    width: ${0}%;
    border-radius: ${0}px;
`),e=>1/e.optionsAmount*100,e=>e.optionBorderRadius),f=n("label")(s||(s=p`
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100%;
    z-index: 2;
    transition: color 0.1s linear;
    cursor: ${0};
    font-size: ${0};
    color: ${0};
`),e=>e.disabled?"not-allowed":"pointer",e=>e.isRawText?e.fontSize+"px":"unset",e=>e.isRawText?e.selected?e.selectedFontColor:e.fontColor:"unset"),C=n("input")(c||(c=p`
    width: 0;
    height: 0;
    opacity: 0;
    z-index: -1;
    position: absolute;
    pointer-events: none;
`)),h={backgroundColor:"#ecf0f1",selectedBackgroundColor:"#2ecc71",fontColor:"#000",selectedFontColor:"#fff"};export default o=>{var r;const{onChange:n=(()=>{}),options:i=[],initialSelectedIndex:l=0}=o,a=!!i[l],[s,c]=e.useState(a?l:0),{border:p=0,backgroundColor:u=h.backgroundColor,selectedBackgroundColor:x=h.selectedBackgroundColor,wrapperBorderRadius:m=20,optionBorderRadius:w=18,fontSize:k=14,fontColor:$=h.fontColor,selectedFontColor:B=h.selectedFontColor,selectionIndicatorMargin:R=2,forcedSelectedIndex:v,disabled:y=!1}=o;return t(()=>{void 0!==v&&i[v]&&v!==s&&c(v)},[v,i,s]),i.length?e.createElement(g,{selectedIndex:s,optionsAmount:i.length,className:"react-switch-selector-wrapper "+(y?"react-switch-selector-disabled":""),border:p,backgroundColor:u,selectedBackgroundColor:(null==(r=i[s])?void 0:r.selectedBackgroundColor)||x,wrapperBorderRadius:m,optionBorderRadius:w,selectionIndicatorMargin:R,disabled:y},i.map((o,t)=>{const r=`rss-option-${t}`,l="string"==typeof o.label;return e.createElement(b,{key:r,optionsAmount:i.length,className:"react-switch-selector-option",optionBorderRadius:w},e.createElement(f,d({className:"react-switch-selector-option-label",selected:s===t,isRawText:l,disabled:y,"aria-disabled":y},l?{fontSize:k,fontColor:o.fontColor||$,selectedFontColor:o.selectedFontColor||B}:{}),e.createElement(C,{type:"radio",id:r,onChange:()=>(e=>{y||e===s||(c(e),n(i[e].value))})(t),checked:s===t}),o.label))})):null};
